bee-message/src/payload/data.rs

```diff
@@ -0,0 +1,159 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidPrefix => write!(f, "invalid prefix for data"),
+            Self::InvalidPrefixLength(len) => write!(f, "unpacked prefix larger than maximum specified: {}", len),
+        }
+    }
+}
+
+/// Generic data payload, containing a collection of bytes.
+///
+/// A [`DataPayload`] must:
+/// * Not exceed [`PAYLOAD_LENGTH_MAX`] in bytes.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
+pub struct DataPayload {
+    /// The raw data in bytes.
+    data: Vec<u8>,
+}
+
+impl MessagePayload for DataPayload {
+    const KIND: u32 = 1;
+    const VERSION: u8 = 0;
+}
+
+impl DataPayload {
+    /// Creates a new [`DataPayload`].
+    pub fn new(data: Vec<u8>) -> Result<Self, ValidationError> {
+        validate_data_len(data.len())?;
+
+        Ok(Self { data })
+    }
+
+    /// Returns the data bytes of a [`DataPayload`].
+    pub fn data(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+}
+
+impl Packable for DataPayload {
+    type PackError = MessagePackError;
+    type UnpackError = MessageUnpackError;
+
+    fn packed_len(&self) -> usize {
+        // Unwrap is safe, since the data length has already been validated.
+        let prefixed_data: VecPrefix<u8, u32, PREFIXED_DATA_LENGTH_MAX> = self.data.clone().try_into().unwrap();
+
+        Self::VERSION.packed_len() + prefixed_data.packed_len()
+    }
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), PackError<Self::PackError, P::Error>> {
+        Self::VERSION.pack(packer).map_err(PackError::infallible)?;
+
+        // Unwrap is safe, since the data length has already been validated.
+        let prefixed_data: VecPrefix<u8, u32, PREFIXED_DATA_LENGTH_MAX> = self.data.clone().try_into().unwrap();
+        prefixed_data
+            .pack(packer)
+            .map_err(PackError::coerce::<DataPackError>)
+            .map_err(PackError::coerce)?;
+
+        Ok(())
+    }
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let version = u8::unpack(unpacker).map_err(UnpackError::infallible)?;
+        validate_payload_version(version).map_err(|e| UnpackError::Packable(e.into()))?;
+
+        let data: Vec<u8> = VecPrefix::<u8, u32, PREFIXED_DATA_LENGTH_MAX>::unpack(unpacker)
+            .map_err(UnpackError::coerce::<DataUnpackError>)
+            .map_err(UnpackError::coerce)?
+            .into();
+        validate_data_len(data.len()).map_err(|e| UnpackError::Packable(e.into()))?;
+
+        let payload = Self { data };
+
+        Ok(payload)
+    }
+}
+
+fn validate_payload_version(version: u8) -> Result<(), ValidationError> {
+    if version != DataPayload::VERSION {
+        Err(ValidationError::InvalidPayloadVersion {
+            version,
+            payload_kind: DataPayload::KIND,
+        })
+    } else {
+        Ok(())
+    }
+}
```

Comment:
what happens if `len == PAYLOAD_LENGTH_MAX`? Wouldn't that cause a problem in line 108 and 117? Are you sure you don't want to check against PREFIXED_DATA_LENGTH_MAX?

Reply:
You're right -- wrong number there.